[SPARK-8496] [TEST] Do not run slow tests for every PR

andrewor14 opened the pull request:
The purpose of this patch is to speed up the development cycle by separating slow tests from normal tests. The intention is to run slow tests only in the non-pull-request-builder Jenkins projects. A slow test is very roughly defined as _a test that takes several seconds to run on our test machines_.

In the future we will provide a way for PRs to specify that they do want to run the full tests.

Commit:
Run slow tests only if they are enabled

Commit:
Mark more tests as slow tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35353/consoleFull) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).

rxin commented:
Can we just use scalatest's tags for this?

andrewor14 commented:
what do you mean?

rxin commented:
It is a feature of ScalaTest -- you can tag individual test cases. And then at runtime you can specify to rule out specific tags, or only run specific tags. See http://www.scalatest.org/user_guide/tagging_your_tests (also available in `test("...")`)

It is basically designed for this purpose, i.e. you can define a LongRunning tag.

SparkQA commented:
[Test build #35353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35353/console) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  slowTest("sort with Java non serializable class - Kryo") `
  * `  slowTest("sort with Java non serializable class - Java") `
  * `  slowTest("Optimize via class LBFGS.") `
  * `  slowTest("SPARK-1199 two instances of same class don't type check.") `
  * `  slowTest("SPARK-2632 importing a method from non serializable class and not using it.") `
  * `  slowTest("collecting objects of class defined in repl") `
  * `  slowTest("collecting objects of class defined in repl - shuffling") `
  * `  slowTest("user class path first in client mode") `
  * `  slowTest("user class path first in cluster mode") `

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35356/consoleFull) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).

SparkQA commented:
[Test build #35356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35356/console) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  slowTest("sort with Java non serializable class - Kryo") `
  * `  slowTest("sort with Java non serializable class - Java") `
  * `  slowTest("Optimize via class LBFGS.") `
  * `  slowTest("SPARK-1199 two instances of same class don't type check.") `
  * `  slowTest("SPARK-2632 importing a method from non serializable class and not using it.") `
  * `  slowTest("collecting objects of class defined in repl") `
  * `  slowTest("collecting objects of class defined in repl - shuffling") `
  * `  slowTest("user class path first in client mode") `
  * `  slowTest("user class path first in cluster mode") `

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35466/consoleFull) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).

harishreedharan commented:
hi @andrewor14 - I agree with @rxin here. We should probably use the framework's tagging mechanism, no?

andrewor14 commented:
@harishreedharan we discussed offline and already agreed on special tags.

harishreedharan commented:
Cool, thank you!

SparkQA commented:
[Test build #35466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35466/console) for   PR 6917 at commit [`9bb1f19`](https://github.com/apache/spark/commit/9bb1f190299feeb40f1f7cdeb64e152b19ef08b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  slowTest("sort with Java non serializable class - Kryo") `
  * `  slowTest("sort with Java non serializable class - Java") `
  * `  slowTest("Optimize via class LBFGS.") `
  * `  slowTest("SPARK-1199 two instances of same class don't type check.") `
  * `  slowTest("SPARK-2632 importing a method from non serializable class and not using it.") `
  * `  slowTest("collecting objects of class defined in repl") `
  * `  slowTest("collecting objects of class defined in repl - shuffling") `
  * `  slowTest("user class path first in client mode") `
  * `  slowTest("user class path first in cluster mode") `

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix tests

These tests actually depend on each other. If we mark the first
one as a slow test, the rest will throw NPE because the first sets
a variable that others need.

Commit:
Merge branch 'master' of github.com:apache/spark into slow-tests

Commit:
Use scalatest tags instead of environment variables

Commit:
Set up tag excludes for SBT and maven

In both cases, the user specifies the environment variable
`SPARK_TEST_EXCLUDED_TAGS` before running `dev/run-tests.py`.
The exclusion logic will then be handled by maven and SBT in
their own respective ways.

Note: we still need to make sure the developer has access to
specific tests on demand. Otherwise if he adds a slow test
he will have no way to make it run in his PR.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35568/consoleFull) for   PR 6917 at commit [`56516ea`](https://github.com/apache/spark/commit/56516ea14c3b2ee89ba134b16d758768583a8b29).

Commit:
Merge branch 'master' of github.com:apache/spark into slow-tests

andrewor14 commented:
Note that we should not merge this PR until we have a way to allow the developer to run specific tests.

AmplabJenkins commented:
Merged build triggered.

SparkQA commented:
[Test build #35568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35568/console) for   PR 6917 at commit [`56516ea`](https://github.com/apache/spark/commit/56516ea14c3b2ee89ba134b16d758768583a8b29).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #35571 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35571/consoleFull) for   PR 6917 at commit [`cdff98f`](https://github.com/apache/spark/commit/cdff98fb8f259f704cde118eba9116426118c103).

SparkQA commented:
[Test build #35571 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35571/console) for   PR 6917 at commit [`cdff98f`](https://github.com/apache/spark/commit/cdff98fb8f259f704cde118eba9116426118c103).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class BroadcastHint(child: LogicalPlan) extends UnaryNode `

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
hey @andrewor14 is this the same as https://issues.apache.org/jira/browse/SPARK-4746?  I dropped this PR b/c it seemed there was no interest: https://github.com/apache/spark/pull/4048 .  I figured out how to get junit "categories" integrated and to get it to work with the sbt test runner as well (it was a little complicated b/c junit & scalatest expect arguments in different incompatible formats).  I think the only thing missing from that was to get it to work w/ maven as well.

My version is certainly extremely out of date by now, but do you want to steal the changes I made to SparkBuild.scala?  I can look into the maven changes as well if that would be helpful

Commit:
Fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
Hi @squito unfortunately JIRA is not loading for me atm, but I suspect it's something similar. This patch already handles both maven and SBT, but only for scalatests. The junit changes in #4048 may be useful.

However, this patch will likely not be merged any time soon because there is no way to enable at least a subset of the disabled tests. Either way I plan to add the junit changes later.

SparkQA commented:
[Test build #35581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35581/consoleFull) for   PR 6917 at commit [`d59759e`](https://github.com/apache/spark/commit/d59759ee14d7ff016365dedb4fee3ad045e36dbd).

SparkQA commented:
[Test build #35581 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35581/console) for   PR 6917 at commit [`d59759e`](https://github.com/apache/spark/commit/d59759ee14d7ff016365dedb4fee3ad045e36dbd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35605/consoleFull) for   PR 6917 at commit [`a707b85`](https://github.com/apache/spark/commit/a707b85de8ba8ad89085fc29f2317dc3dfee626a).

SparkQA commented:
[Test build #35605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35605/console) for   PR 6917 at commit [`a707b85`](https://github.com/apache/spark/commit/a707b85de8ba8ad89085fc29f2317dc3dfee626a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
This patch in its current state reduces the time roughly from `150m` to `104m`. This does not currently optimize python or java tests.

squito commented on dev/run-tests.py:
```diff
@@ -639,32 +651,47 @@ def detect_binary_inop_with_mima():
+    if tags_to_exclude:
+        new_sbt_test_goals = []
+        for g in sbt_test_goals:
+            g = re.sub(r"test$", "test-only * -- -l %s" % tags_to_exclude, g)
```
I think this will have the unintended side-effect of excluding *all* junit tests.  Its really unfortunate that the junit args & the scalatest args are incompatible.  From https://github.com/sbt/junit-interface, "Any parameter not starting with - or + is treated as a glob pattern for matching tests.", so junit decides to only look for tests which match the pattern of the test you are trying to exclude.  Eg., compare:

```
> test-only *JavaAPISuite
...
[info] Test run finished: 0 failed, 0 ignored, 90 total, 20.878s
```

vs

```
> test-only *JavaAPISuite -- -l FooBar
...
[info] Test run finished: 0 failed, 0 ignored, 0 total, 0.001s
```


That's why I had to make bigger changes to the sbt build to get it to pass a different set of args to junit & scalatest

andrewor14 replied:
That's too bad. It seems that `SparkSubmitOptionParserSuite.java` did run, however. Do you know why?

squito replied:
@andrewor14 sorry I missed this comment earlier.  I'm not sure why.  It seems it only ran junit tests for the launcher project, but I don't see anything special about that one project in particular ...

JoshRosen commented:
I think that we should put this on hold for now and revisit after we've refactored the Jenkins triggering script so that we can explicitly tell Jenkins whether to run full tests or not.

andrewor14 commented:
I'm going to close this until we can do that. Note to self: DO NOT delete the branch.

SparkQA commented:
[Test build #37154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37154/consoleFull) for   PR 6917 at commit [`a707b85`](https://github.com/apache/spark/commit/a707b85de8ba8ad89085fc29f2317dc3dfee626a).

SparkQA commented:
[Test build #37154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37154/console) for   PR 6917 at commit [`a707b85`](https://github.com/apache/spark/commit/a707b85de8ba8ad89085fc29f2317dc3dfee626a).
 * This patch **fails MiMa tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

squito commented:
Hey @andrewor14 @JoshRosen what about merging something like this *without* the changes to the PR builder?  then at least the tests are tagged, so (a) developers can actually run a reasonable subset of tests locally (sbt's `~test-quick` is really great, if you can wait for the first run ...) and (b) it'll be easier to make sure future tests get properly labeled?
